### PR TITLE
Workflow editor improvements - add node from empty connection and auto-connect to empy handle.

### PR DESCRIPTION
**Pull request by CrypticWit**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [ ] Yes
- [x] No

## Description

Implemented creation of nodes by dragging a new connection from a handle to an empty space on the canvas.  The Add Node popover will only show valid nodes that can connect to the initial handle.

You can also drag an edge onto an existing node to have it connect to the first valid handle that is not currently connected.  This allows for faster (and less precise) connection between existing nodes.

The following hasn't been fully implemented yet or I need some advice on:
- New nodes should probably be created at the mouse cursor, rather than the center of the workspace
- dragging a connection onto an existing node can be a bit fiddly at times due to fields wanting to take over input. Can we override the default behaviour?
- I'm not sure if I'm creating the edges/checking for valid handles in the correct 'React'ish way.  If there is a better way of doing this, I'm happy to refactor things.

## Added/updated tests?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by Millu** [comment]
I love this - such a nice QoL improvement. The only pain point with usage is the edge connects to the first match input on the new node which is not always desired. Not sure if it makes sense to change that, but because you can't move an edge to another input / output once it's been connected it feels cumbersome. 

Adding the ability to click and drag an edge from a previously connected input to a new one would fix this. Whether that's for this PR or another I'm not sure

Thank you for putting this together @CrypticWit!

**Comment by CrypticWit** [comment]
> I love this - such a nice QoL improvement. The only pain point with usage is the edge connects to the first match input on the new node which is not always desired. Not sure if it makes sense to change that, but because you can't move an edge to another input / output once it's been connected it feels cumbersome.
> 
> Adding the ability to click and drag an edge from a previously connected input to a new one would fix this. Whether that's for this PR or another I'm not sure
> 
> Thank you for putting this together @CrypticWit!

I can take a look tonight and see what might be required for that, unless @psychedelicious knows off the top of his head how much effort this would be.

Something else for a future PR to improve this experience going forward is having a way to search the node graph a little more extensively to further filter the Add node options.  if you're dragging a condition handle from a  a denoise latent node that is already connected to a SDXL Unet, you don't want to see the base SD prompt node, but instead only the SDXL compatible nodes.   Whether this is functionality that should be handled in Add Node or by making some nodes a little more polymorphic for model types is probably a bigger conversation though.

**Comment by psychedelicious** [comment]
I've implemented updatable edges in  #4701. 

I think this PR will need to work around it slightly - when you drop an edge away from a node, the edge is deleted. That may conflict with the behaviour in the PR.

Perhaps the mouseup on empty connection behaviour (delete edge vs add node) should have a toggle?

**Comment by Millu** [comment]
> Perhaps the mouseup on empty connection behaviour (delete edge vs add node) should have a toggle?

The behavior you have in #4701 makes sense to me - it's the same way Blender does it. 

So expected behavior is the add node functionality appears from an edge only if it's a new edge

**Comment by CrypticWit** [comment]
Looking at the PR, it should be easy enough to just add a flag to not show the add node popup if edgeDelete was called before connectionEnded - if it was, then this is an old node, otherwise its a new edge, and so we want to show the popup?

**Comment by CrypticWit** [comment]
I have implemented this behaviour, and it works well. The only issue remaining is that there is a very small difference between clicking on the handle itself, and clicking on the 'draggable' section of the edge - and if you click the handle by accident, you're left with a new edge that will spawn the popover when you release it.

The only solution I can think of at the moment to this is to have a minimum edge distance for the popover to appear - but then we'll need to decide what distance that should be (and probably in screen-space, rather than projected space)

**Comment by hipsterusername** [comment]
@CrypticWit - in my mind, distance is probably roughly quarter the width of a node, and this can be communicated through some form of styling on the edge that’s being dragged. 

Definitely some ways to have this make things annoying though (e.g., if it disabled the ability to connect to a valid node that was immediately adjacent)

— BTW - awesome QOL enhancement, love it!

File: invokeai/frontend/web/src/features/nodes/store/nodesSlice.ts [hunk 1]
```diff
@@ -835,10 +915,15 @@ const nodesSlice = createSlice({
...
     addNodePopoverOpened: (state) => {
+      state.addNewNodePosition = null; //Create the node in viewport center by default
       state.isAddNodePopoverOpen = true;
     },
     addNodePopoverClosed: (state) => {
       state.isAddNodePopoverOpen = false;
+
+      //Make sure these get reset if we close the popover and haven't selected a node
+      state.currentConnectionFieldType = null;
+      state.currentConnectionFieldType = null;
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/nodes/store/nodesSlice.ts:926]: Setting same value twice?

Comment (CrypticWit) [reply invokeai/frontend/web/src/features/nodes/store/nodesSlice.ts]: Fixed

**Comment by CrypticWit** [comment]
I think leaving the behaviour as it is now (cancel the line by dropping it back on the node it came from), is probably the way to go. I can prototyping up some ideas once more people get to play with it if it becomes an actual issue.

**Comment by Millu** [comment]
Just gave this a whirl again and happy with the behavior as is. Let's get this in as is, and if the spacing between the node vertex and the edge proves difficult for users, we can revisit this!

**Review by Millu — APPROVED** [review]

**Comment by psychedelicious** [comment]
I made two changes:

1. [fix(ui): do not allow Collect -> Iterate connection](https://github.com/invoke-ai/InvokeAI/pull/4684/commits/8e58c8a572fefc9917cb1dc90d98a441fc796aed) This can be removed when https://github.com/invoke-ai/InvokeAI/issues/3956 is resolved

2. [feat(ui): use existing node validation logic in add-node-on-drop](https://github.com/invoke-ai/InvokeAI/pull/4684/commits/d6109a826fd04bbba632b521a7131a3c64004e04) This logic handles a number of special cases

This handles things like "collections of a type are allowed to connect to both collections or polymorphics of that type".

I used some `lodash` utils to clean up the filtering logic a bit. I encourage `lodash` for this project because it generally results in cleaner, easier-to-grok code. Note that you must import from `lodash-es` to get tree-shaking.

**Review by psychedelicious — APPROVED** [review]
Thanks! This is a great QoL improvement, works wonderfully.

**Pull request merged**